### PR TITLE
[docs] optimizers, hyperparam search, training features

### PR DESCRIPTION
**Pull request by stevhliu**
- adds `trainer_recipes.md` to show how to use other practical `Trainer` features outside of the basic training loop
- updates hyperparam search docs
- updates `optimizers.md` with how to customize it (prebuilt instances, passing a class + kwargs, subclassing)

### CODE REVIEW

File: docs/source/en/optimizers.md [hunk 1]
```diff
@@ -14,63 +14,100 @@ rendered properly in your Markdown viewer.
...
+  │              optimizer.step() ◄── reads lr ◄──────────────────┘   │
+  │                          │                                        │
+  │                          ▼                                        │
+  │                   param.data updated                              │
+  │                          │                                        │
+  │                          ▼                                        │
+  │              lr_scheduler.step()  ──► recalculates lr             │
+  │                          │            writes to optimizer         │
+  │                          ▼            .param_groups['lr']         │
+  │              optimizer.zero_grad()                                │
```
Comment (SunMarc) [inline docs/source/en/optimizers.md]: model.zero_grad()

File: docs/source/en/optimizers.md [hunk 2]
```diff
@@ -14,63 +14,100 @@ rendered properly in your Markdown viewer.
...
+
+```md
+                                    ┌────────────┐         ┌──────────────┐
+                                    │ Optimizer  │         │  Scheduler   │
+                                    │ (adamw_torch_fused)◄─│  (linear)    │
+                                    │            │         │              │
+                                    │ param_groups         │ lr_lambda()  │
+                                    │  └ lr       ◄────────┤ step counter │
+                                    │  └ weight_decay      │              │
+                                    └──────┬─────┘         └──────┬───────┘
```
Comment (SunMarc) [inline docs/source/en/optimizers.md]: i think we don't need to specify  lr_lambda() or step counter maybe ? Maybe we can just have an arrow to lr to show that it modifies the lr.

File: docs/source/en/optimizers.md [hunk 3]
```diff
@@ -14,63 +14,100 @@ rendered properly in your Markdown viewer.
...
+  ┌──── EACH TRAINING STEP ───────────────────────────────────────────┐
+  │                                        │                      │   │
+  │   model(batch)                         │                      │   │
+  │       │                                │                      │   │
+  │       ▼                                │                      │   │
+  │     loss ──► loss.backward() ──► param.grad                   │   │
+  │                                        │                      │   │
+  │                          ┌─────────────┘                      │   │
+  │                          ▼                                    │   │
+  │              optimizer.step() ◄── reads lr ◄──────────────────┘   │
```
Comment (SunMarc) [inline docs/source/en/optimizers.md]: we don't need `◄── reads lr ` I think . The optimizer just uses the lr value that is stored in the optimizer. the scheduler do nothing here

File: docs/source/en/optimizers.md [hunk 4]
```diff
@@ -14,63 +14,100 @@ rendered properly in your Markdown viewer.
...
     learning_rate=2e-5,
-    save_strategy="no",
-    run_name="optimizer-name",
+    weight_decay=0.01,
+    adam_beta1=0.9,
+    adam_beta2=0.999,
+    adam_epsilon=1e-8,
+    # Scheduler
+    lr_scheduler_type="cosine",   # "linear", "cosine", "constant_with_warmup", etc.
+    warmup_steps=500,             # or warmup_ratio=0.06 (fraction of total steps)
```
Comment (SunMarc) [inline docs/source/en/optimizers.md]: `warmup_ratio` don't exist anymore, you can set directly warmup_steps=0.6

```suggestion
    warmup_steps=500,
```

File: docs/source/en/optimizers.md [hunk 5]
```diff
@@ -14,63 +14,100 @@ rendered properly in your Markdown viewer.
...
+Transformers integrates third-party optimizers for specialized training scenarios.
+
+| Optimizer | Install | `optim="value"` | Description |
+|---|---|---|---|
+| APOLLO | `apollo-torch` | `apollo_adamw` | Memory-efficient full-param via random projections; rank-1 sufficient |
+| GrokAdamW | `grokadamw` | `grokadamw` | Targets delayed generalization (grokking) |
+| LOMO / AdaLomo | `lomo-optim` | `lomo` / `adalomo` | Fuses gradient + update step for low-memory full-param fine-tuning |
+| Schedule Free | `schedulefree` | `schedule_free_adamw`, `schedule_free_radam`, `schedule_free_sgd` | Eliminates LR annealing; pair with `lr_scheduler_type="constant"` |
+| GaLore | `galore-torch` | `galore_adamw`, `galore_adafactor`, `galore_adamw_8bit` | Full-parameter learning via gradient low-rank projection |
+| StableAdamW | `torch-optimi` | `stable_adamw` | AdamW + AdaFactor update clipping; no gradient clipping needed |
```
Comment (SunMarc) [inline docs/source/en/optimizers.md:110]: Thanks ! This is much better

File: docs/source/en/optimizers.md [hunk 6]
```diff
@@ -80,196 +117,228 @@ For additional training options, use `optim_args` to define hyperparameters like
...
 ```diff
 args = TrainingArguments(
-    output_dir="./test-lomo",
-    max_steps=1000,
-    per_device_train_batch_size=4,
 +   optim="adalomo",
-    gradient_checkpointing=True,
-    logging_strategy="steps",
-    logging_steps=1,
++   gradient_checkpointing=True,
```
Comment (SunMarc) [inline docs/source/en/optimizers.md]: let's remove the gradient_checkpointing since there is no specific description about this with this optimizer

File: docs/source/en/optimizers.md [hunk 7]
```diff
@@ -80,196 +117,228 @@ For additional training options, use `optim_args` to define hyperparameters like
...
-    output_dir="./test-schedulefree",
-    max_steps=1000,
-    per_device_train_batch_size=4,
-+   optim="schedule_free_radamw",
++   optim="schedule_free_radam",
 +   lr_scheduler_type="constant",
-    gradient_checkpointing=True,
-    logging_strategy="steps",
-    logging_steps=1,
++   gradient_checkpointing=True,
```
Comment (SunMarc) [inline docs/source/en/optimizers.md]: same

File: docs/source/en/optimizers.md [hunk 8]
```diff
@@ -80,196 +117,228 @@ For additional training options, use `optim_args` to define hyperparameters like
...
 ```diff
 args = TrainingArguments(
-    output_dir="./test-stable-adamw",
-    max_steps=1000,
-    per_device_train_batch_size=4,
 +   optim="stable_adamw",
-    gradient_checkpointing=True,
-    logging_strategy="steps",
-    logging_steps=1,
++   gradient_checkpointing=True,
```
Comment (SunMarc) [inline docs/source/en/optimizers.md]: rm

File: docs/source/en/optimizers.md [hunk 9]
```diff
@@ -80,196 +117,228 @@ For additional training options, use `optim_args` to define hyperparameters like
...
 
-train_dataset = datasets.load_dataset('imdb', split='train')
 args = SFTConfig(
-    output_dir="./test-galore",
+    output_dir="./galore",
     max_steps=100,
     optim="galore_adamw",
     optim_target_modules=[r".*.attn.*", r".*.mlp.*"],
     optim_args="rank=64, update_proj_gap=100, scale=0.10",
     gradient_checkpointing=True,
```
Comment (SunMarc) [inline docs/source/en/optimizers.md]: also

File: docs/source/en/optimizers.md [hunk 10]
```diff
@@ -80,196 +117,228 @@ For additional training options, use `optim_args` to define hyperparameters like
...
 
-train_dataset = datasets.load_dataset('imdb', split='train')
 args = SFTConfig(
-    output_dir="./test-galore",
+    output_dir="./galore",
     max_steps=100,
     optim="galore_adamw_layerwise",
     optim_target_modules=[r".*.attn.*", r".*.mlp.*"],
     optim_args="rank=64, update_proj_gap=100, scale=0.10",
     gradient_checkpointing=True,
```
Comment (SunMarc) [inline docs/source/en/optimizers.md]: here also

File: docs/source/en/optimizers.md [hunk 11]
```diff
@@ -80,196 +117,228 @@ For additional training options, use `optim_args` to define hyperparameters like
...
+trainer = Trainer(
+    ...
+    optimizer_cls_and_kwargs=(
+        torch.optim.SGD,
+        {"momentum": 0.9, "nesterov": True}
+    ),
+)
+```
+
+[`~Trainer.optimizer_cls_and_kwargs`] doesn't allow a custom scheduler and can't be combined with [`~Trainer.optimizers`].
```
Comment (SunMarc) [inline docs/source/en/optimizers.md]: not needed, create some confusion

File: docs/source/en/optimizers.md [hunk 12]
```diff
@@ -80,196 +117,228 @@ For additional training options, use `optim_args` to define hyperparameters like
...
+import torch
+
+trainer = Trainer(
+    ...
+    optimizer_cls_and_kwargs=(
+        torch.optim.SGD,
+        {"momentum": 0.9, "nesterov": True}
+    ),
+)
+```
```
Comment (SunMarc) [inline docs/source/en/optimizers.md]: this is the prefered path, so i recommend putting that above. For the other one, it feels simpler but as you saw, the model needs to be initialized correctly before creating the optimizer

File: docs/source/en/optimizers.md [hunk 13]
```diff
@@ -80,196 +117,228 @@ For additional training options, use `optim_args` to define hyperparameters like
...
+### Override optimizer and scheduler methods
+
+Subclass [`~Trainer.create_optimizer`] and [`~Trainer.create_scheduler`] for full control. Both methods run *during* [`~Trainer.train`].
+
+Override [`~Trainer.create_scheduler`] to use a scheduler like [OneCycleLR](https://docs.pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.OneCycleLR.html) that isn't available in [`SchedulerType`].
+
+For each method, make sure to assign to `self` and return it.
+
+1. Assign `OneCycleLR` to `self.lr_scheduler`.
+2. Set `self._created_lr_scheduler = True`, otherwise [`Trainer`] rebuilds the scheduler and overwrites `OneCycleLR`.
```
Comment (SunMarc) [inline docs/source/en/optimizers.md]: this is like a trick for auto_batch_size to work correctly. let's not confuse the reader and remove this I think

File: docs/source/en/trainer_recipes.md [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+Pass [compute_loss_func](https://huggingface.co/docs/transformers/en/main_classes/trainer#transformers.Trainer.compute_loss_func) to [`Trainer`] to replace the default loss function. The function runs *after* the forward pass and only defines how loss is computed from the outputs. To modify the forward pass itself, [subclass](./trainer_customize#compute_loss) [`~Trainer.compute_loss`] instead.
+
+The custom loss function must have the following signature:
+
+```py
+import torch
+import torch.nn.functional as F
+from transformers import Trainer, TrainingArguments
+
+def my_loss_fn(outputs, labels, num_items_in_batch=None):
```
Comment (SunMarc) [inline docs/source/en/trainer_recipes.md]: `num_items_in_batch` shouldn't be optional, otherwise they will get weird results when using gradient acc or training with multiple gpus

File: docs/source/en/trainer_recipes.md [hunk 15]
```diff
@@ -0,0 +1,255 @@
...
+    logits = outputs["logits"]
+    if num_items_in_batch is not None:
+        loss = F.cross_entropy(logits, labels, reduction="sum")
+        loss = loss / num_items_in_batch
+    else:
+        loss = F.cross_entropy(logits, labels, reduction="mean")
```
Comment (SunMarc) [inline docs/source/en/trainer_recipes.md]:
```suggestion
    loss = F.cross_entropy(logits, labels, reduction="sum")
    loss = loss / num_items_in_batch
```

File: docs/source/en/trainer_recipes.md [hunk 16]
```diff
@@ -0,0 +1,255 @@
+<!--Copyright 2026 The HuggingFace Team. All rights reserved.
```
Comment (SunMarc) [inline docs/source/en/trainer_recipes.md:1]: we can keep these here for now, but i feel like they can be moved in other sections.  Trainer receipe name don't fully represent what we have here. Maybe we can think of another name ? I feel like we could create another kind of recipes here.

**Review by SunMarc — COMMENTED** [review]
Thanks for these, this is already much better, left a couple of comments

File: docs/source/en/trainer_recipes.md (hunk 16 above)
Comment (stevhliu) [reply docs/source/en/trainer_recipes.md:1]: hmm yeah a recipe is more of a contained end-to-end pattern. i'll rename this to "Trainer features" which i think is more accurate of what is described here. this would free up Trainer recipes for actual recipes later

**Comment by stevhliu** [comment]
i updated the scheduler docs to include `GreedyLR` from #44271 (sorry i missed your comment!)

**Comment by stevhliu** [comment]
updated with flashoptim from #44438!

**Review by SunMarc — APPROVED** [review]
Thanks a lot !
